sled-agent/src/rack_setup/service.rs

```diff
@@ -641,35 +660,88 @@ impl ServiceInner {
                             };
                             internal_services_ip_pool_ranges.push(range);
 
-                            NexusTypes::ServiceKind::Nexus {
-                                external_address: *external_ip,
-                            }
+                            services.push(NexusTypes::ServicePutRequest {
+                                service_id: zone.id,
+                                sled_id,
+                                address: SocketAddrV6::new(
+                                    zone.addresses[0],
+                                    NEXUS_INTERNAL_PORT,
+                                    0,
+                                    0,
+                                )
+                                .to_string(),
+                                kind: NexusTypes::ServiceKind::Nexus {
+                                    external_address: *external_ip,
+                                },
+                            });
                         }
-                        ServiceType::InternalDns { .. } => {
-                            NexusTypes::ServiceKind::InternalDNS
+                        ServiceType::InternalDns {
+                            server_address,
+                            dns_address,
+                        } => {
+                            services.push(NexusTypes::ServicePutRequest {
+                                service_id: zone.id,
+                                sled_id,
+                                address: server_address.to_string(),
+                                kind:
+                                    NexusTypes::ServiceKind::InternalDNSConfig,
+                            });
+                            services.push(NexusTypes::ServicePutRequest {
+                                service_id: zone.id,
+                                sled_id,
+                                address: dns_address.to_string(),
+                                kind: NexusTypes::ServiceKind::InternalDNS,
+                            });
```

Comment:
This was in the back of my mind and I had forgotten but I don't think we should be using `zone.id` as the `service_id` in `ServicePutRequest`. Any subsequent entries with the same id will just overwrite the previous ones:

https://github.com/oxidecomputer/omicron/blob/5d39727039cca86a7fef48d35f3675187386c333/nexus/db-queries/src/db/datastore/rack.rs#L181-L193

This wasn't previously an issue since we only ever had 1 service per zone up til this point AFAIK.

And sorry for the post-merge comments here, I had meant to look but earlier :/

Reply:
Thank you for catching this!

I wonder if we should keep calling this `service_id` (or `zone_id`) but make the primary key (`service_id`, `kind`).  My thinking here is that It'd be annoying to have some extra uuid that's not traceable to anything in particular, and also to have no way to get to the corresponding zone id.

Reply:
A tuple primary key w/ `kind` would fix the current setup but not if we want to ever run two instances of the same service in the same zone. (Not saying we'd ever need/want to but there's no reason we couldn't). Maybe we just add a ` zone_id` column? A little verbose but makes things super clear

Reply:
Filed as #2813.